ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/io/BigQuery.java

```diff
@@ -0,0 +1,124 @@
+
+    public Write(com.google.cloud.bigquery.BigQuery bigquery) {
+      this.bigquery = bigquery;
+    }
+
+    @Override
+    public CompletableFuture<TableRow> apply(TableRow row) {
+      return Optional.ofNullable(batches.get(row.tableId)).flatMap(batch -> batch.add(row))
+          .orElseGet(() -> {
+            final Batch batch = new Batch(bigquery, row.tableId);
+            final CompletableFuture<TableRow> response = batch.add(row)
+                .orElseThrow(() -> new IllegalArgumentException("Empty batch rejected message"));
+            batches.put(row.tableId, batch);
+            return response;
+          });
+    }
+
+    /**
+     * POJO of the information necessary to insert a row into BigQuery.
+     */
+    public static class TableRow {
+
+      public final TableId tableId;
+      public final int byteSize;
+      public final Map<String, Object> content;
+
+      public TableRow(TableId tableId, int byteSize, Map<String, Object> content) {
+        this.tableId = tableId;
+        this.byteSize = byteSize;
+        this.content = content;
+      }
+    }
+
+    private class WriteErrors extends RuntimeException {
+
+      public final List<BigQueryError> errors;
+
+      private WriteErrors(List<BigQueryError> errors) {
+        this.errors = errors;
+      }
+    }
+
+    private class Batch {
+
+      private final com.google.cloud.bigquery.BigQuery bigquery;
+      private final InsertAllRequest.Builder builder;
+      private final CompletableFuture<Void> full;
+      private final CompletableFuture<InsertAllResponse> result;
+
+      int size = 0;
+      int byteSize = 0;
+
+      private Batch(com.google.cloud.bigquery.BigQuery bigquery, TableId tableId) {
+        this.bigquery = bigquery;
+        builder = InsertAllRequest.newBuilder(tableId)
+            // ignore row values for columns not present in the table
+            .setIgnoreUnknownValues(true)
+            // insert all valid rows when invalid rows are present in the request
+            .setSkipInvalidRows(true);
+        full = CompletableFuture.runAsync(this::timeout);
+        result = full.handleAsync(this::insertAll);
+      }
+
+      private void timeout() {
+        try {
+          Thread.sleep(maxDelayMillis);
+        } catch (InterruptedException e) {
+          // this is fine
+        }
+      }
+
+      private synchronized InsertAllResponse insertAll(Void v, Throwable t) {
+        return bigquery.insertAll(builder.build());
+      }
+
+      private synchronized Optional<CompletableFuture<TableRow>> add(TableRow row) {
+        if (full.isDone()) {
+          return Optional.empty();
+        }
+        if (size > 0 && (size + 1 > maxMessages || byteSize + row.byteSize + 1 > maxBytes)) {
+          this.full.complete(null);
+          return Optional.empty();
+        }
+        size += 1;
+        byteSize += row.byteSize + 1;
+        builder.addRow(row.content);
```

Comment:
I'd advocate that we pull out `document_id` if available and set it as the message ID here so that the BigQuery streaming service can deduplicate if we double-publish.